src/charm_config.py

```diff
@@ -0,0 +1,155 @@
+
+    bridge = "bridge"
+    macvlan = "macvlan"
+    host_device = "host-device"
+    vfioveth = "vfioveth"
+
+
+class UpfMode(str, Enum):
+    """Class to define available UPF modes for UPF operator."""
+
+    af_packet = "af_packet"
+    dpdk = "dpdk"
+
+
+NetworkType: TypeAlias = "str | bytes | int | tuple[str | bytes | int, str | int]"
+
+
+class LaxIPvAnyNetwork(IPvAnyNetwork):
+    """Validate an IPv4 or IPv6 network."""
+
+    def __new__(cls, value: NetworkType):
+        """Validate an IPv4 or IPv6 network."""
+        # Assume IP Network is defined with a default value for `strict` argument.
+        # Define your own class if you want to specify network address check strictness.
+        try:
+            return IPv4Network(value, strict=False)
+        except ValueError:
+            pass
+
+        try:
+            return IPv6Network(value, strict=False)
+        except ValueError:
+            raise PydanticCustomError(
+                "ip_any_network", "value is not a valid IPv4 or IPv6 network"
+            )
+
+
+class CharmConfigInvalidError(Exception):
+    """Exception raised when a charm configuration is found to be invalid."""
+
+    def __init__(self, msg: str):
+        """Initialize a new instance of the CharmConfigInvalidError exception.
+
+        Args:
+            msg (str): Explanation of the error.
+        """
+        self.msg = msg
+
+
+def to_kebab(name: str) -> str:
+    """Converts a snake_case string to kebab-case."""
+    return name.replace("_", "-")
+
+
+class UpfConfig(BaseModel):  # pylint: disable=too-few-public-methods
+    """Represent UPF operator builtin configuration values."""
+
+    model_config = ConfigDict(alias_generator=to_kebab, use_enum_values=True)
+
+    cni_type: CNIType = CNIType.bridge
+    upf_mode: UpfMode = UpfMode.af_packet
+    dnn: StrictStr = Field(default="internet", min_length=1)
+    gnb_subnet: LaxIPvAnyNetwork = LaxIPvAnyNetwork("192.168.251.0/24")
+    access_interface: Optional[StrictStr] = Field(default="")
+    access_interface_mac_address: MacAddress = Field(default=None)
+    access_ip: LaxIPvAnyNetwork = LaxIPvAnyNetwork("192.168.252.3/24")
+    access_gateway_ip: IPvAnyAddress = IPvAnyAddress("192.168.252.1")
+    access_interface_mtu_size: Optional[int] = Field(
+        default=None, ge=1200, le=65535, validate_default=True
+    )
+    core_interface: Optional[StrictStr] = Field(default="")
+    core_interface_mac_address: MacAddress = Field(default=None)
+    core_ip: LaxIPvAnyNetwork = LaxIPvAnyNetwork("192.168.250.3/24")
+    core_gateway_ip: IPvAnyAddress = IPvAnyAddress("192.168.250.1")
+    core_interface_mtu_size: Optional[int] = Field(default=None, ge=1200, le=65535)
+    external_upf_hostname: Optional[StrictStr] = Field(default="")
+    vlan_id: Optional[int] = Field(default=None, ge=1, le=4095)
+    enable_hw_checksum: bool = True
+
+    @model_validator(mode="after")
+    @classmethod
+    def validate_upf_mode_with_mac_addresses(cls, values):
```

Comment:
according to https://docs.pydantic.dev/latest/concepts/validators/

for `mode='after'` validators You should not use (cls, ModelType) as the signature, instead just use (self)